test_util/src/testenv.rs

```diff
@@ -0,0 +1,342 @@
+            );
+        }
+        self
+    }
+
+    /// Reference to the path of the temporary directory.
+    pub fn temp_path(&self) -> &Path {
+        &self.dir_path
+    }
+
+    /// Reference to the working directory the test will be run from.
+    pub fn work_dir(&self) -> &Path {
+        &self.work_dir
+    }
+
+    /// Reference to environment variables that will be set **before** the test.
+    pub const fn env_vars(&self) -> &EnvMap {
+        &self.env_vars
+    }
+
+    /// Get a reference to the environment files that will created
+    pub fn envfiles(&self) -> &[EnvFile] {
+        &self.envfiles
+    }
+
+    fn add_envfile_assume_valid(&mut self, path: PathBuf, contents: Vec<u8>) -> &mut Self {
+        let envfile = EnvFile { path, contents };
+        self.envfiles.push(envfile);
+        self
+    }
+
+    fn assert_envfile_path_is_valid(&self, path: &Path) {
+        assert!(
+            path != self.temp_path(),
+            "path cannot be empty or the same as the temporary directory"
+        );
+        assert!(
+            !self.envfiles.iter().any(|f| f.path == path),
+            "envfile already in testenv: {}",
+            path.display()
+        );
+    }
+
+    fn assert_env_var_is_valid(&self, key: &str) {
+        assert!(!key.is_empty(), "key cannot be empty");
+        assert!(
+            !self.env_vars.contains_key(key),
+            "key already in testenv: {key}"
+        );
+    }
+}
+
+impl Default for TestEnv {
+    fn default() -> Self {
+        let mut testenv = Self::init();
+        testenv.add_env_var(DEFAULT_EXISTING_KEY, DEFAULT_EXISTING_VALUE);
+        testenv.add_envfile(".env", create_default_envfile());
+        testenv
+    }
+}
+
+/// Get a guarded copy of the original process' env vars.
+fn get_env_locker() -> Arc<Mutex<EnvMap>> {
+    Arc::clone(ENV_LOCKER.get_or_init(|| {
+        let map: EnvMap = env::vars().collect();
+        Arc::new(Mutex::new(map))
+    }))
+}
+
+/// Reset the process' env vars back to what was in `original_env`.
+fn reset_env(original_env: &EnvMap) {
+    // remove keys if they weren't in the original environment
+    env::vars()
+        .filter(|(key, _)| !original_env.contains_key(key))
+        .for_each(|(key, _)| env::remove_var(key));
+    // ensure original keys have their original values
+    original_env
+        .iter()
+        .for_each(|(key, value)| env::set_var(key, value));
+}
+
+/// Create an environment to run tests in.
+///
+/// Writes the envfiles, sets the working directory, and sets environment vars.
+fn create_env(testenv: &TestEnv) {
+    env::set_current_dir(&testenv.work_dir).expect("setting working directory");
+
+    for EnvFile { path, contents } in &testenv.envfiles {
+        create_envfile(path, contents);
+    }
+
+    for (key, value) in &testenv.env_vars {
+        env::set_var(key, value);
+    }
+}
+
+/// Create an envfile for use in tests.
+fn create_envfile(path: &Path, contents: &[u8]) {
```

Comment:
Usage of the underscore in `env_file` should be consistent. I think `env_file` is better than without the underscore, to match `env_var`.

Reply:
Good catch